integration-libs/opf/base/root/services/opf-resource-loader.service.ts

```diff
@@ -0,0 +1,198 @@
+    };
+
+    if (resource.attributes) {
+      resource.attributes.forEach((attribute) => {
+        attributes[attribute.key] = attribute.value;
+      });
+    }
+
+    if (resource.url && !this.hasScript(resource.url)) {
+      super.embedScript({
+        src: resource.url,
+        attributes: attributes,
+        callback: () => {
+          this.markResourceAsLoaded(resource, resources, resolve);
+        },
+        errorCallback: () => {
+          this.handleLoadingResourceError(resource.url);
+        },
+      });
+    } else {
+      this.markResourceAsLoaded(resource, resources, resolve);
+    }
+  }
+
+  protected loadStyles(
+    resource: OpfDynamicScriptResource,
+    resources: OpfDynamicScriptResource[],
+    resolve: (value: void | PromiseLike<void>) => void
+  ) {
+    if (resource.url && !this.hasStyles(resource.url)) {
+      this.embedStyles({
+        src: resource.url,
+        callback: () => this.markResourceAsLoaded(resource, resources, resolve),
+        errorCallback: () => {
+          this.handleLoadingResourceError(resource.url);
+        },
+      });
+    } else {
+      this.markResourceAsLoaded(resource, resources, resolve);
+    }
+  }
+
+  executeScriptFromHtml(html: string | undefined) {
+    if (html) {
+      const element = new DOMParser().parseFromString(html, 'text/html');
+      const script = element.getElementsByTagName('script');
+      if (!script?.[0]?.innerText) {
+        return;
+      }
+      Function(script[0].innerText)();
+    }
+  }
+
+  clearAllProviderResources() {
+    this.document
+      .querySelectorAll(`[${this.OPF_RESOURCE_ATTRIBUTE_KEY}]`)
+      .forEach((resource: undefined | HTMLLinkElement | HTMLScriptElement) => {
+        if (resource) {
+          resource.remove();
+        }
+      });
+  }
+
+  loadProviderResources(
+    scripts: OpfDynamicScriptResource[] = [],
+    styles: OpfDynamicScriptResource[] = []
+  ): Promise<void> {
+    const resources: OpfDynamicScriptResource[] = [
+      ...scripts.map((script) => ({
+        ...script,
+        type: OpfDynamicScriptResourceType.SCRIPT,
+      })),
+      ...styles.map((style) => ({
+        ...style,
+        type: OpfDynamicScriptResourceType.STYLES,
+      })),
+    ];
+    if (!resources.length) {
+      return Promise.resolve();
+    }
+    return new Promise((resolve) => {
+      this.loadedResources = [];
+
+      resources.forEach((resource: OpfDynamicScriptResource) => {
+        if (!resource.url) {
+          this.markResourceAsLoaded(resource, resources, resolve);
+        } else {
+          switch (resource.type) {
+            case OpfDynamicScriptResourceType.SCRIPT:
+              this.loadScript(resource, resources, resolve);
+              break;
+            case OpfDynamicScriptResourceType.STYLES:
+              this.loadStyles(resource, resources, resolve);
+              break;
+            default:
+              break;
+          }
+        }
+      });
+    });
+  }
```

Comment:
[MEDIUM]

It is difficult to reason about the flow of data here, especially regarding where is the logic is responsible for calling the `resolve()` callback of the initial promise. Currently, we pass this `resolve` callback to multiple methods, and each of them is responsible for calling the `resolve` callback when all resources are loaded. IMHO it is too much responsibility for those multiple method invocations. We should invert the responsibility. It is those method invocations that should return a their own Promise. And in the main logic of `loadProviderResources()` we should collect all those promises and call on them `Promise.allSettled(resourcePromises)`. 

Side note, unrelated: As far as I understand, the `clearAllProviderResources` method only removes the reouserces from DOM, but doesn't empty the state of the array of `loadedResources` (i.e. the array `loadedResources` still has the same length after calling `clearAllProviderResources`). I guess this is bug-prone. BUT I believe that when we simplify the flow of promises described in he paragraph above, we will be able to remove the complexity of having all those things: `loadedResources` array, `isResourceLoadingCompleted` method, and the `markResourceAsLoaded` method.

Reply:
[OUTDATED] Done.